enhancements/microshift/microshift-apiserver-custom-certs.md

```diff
@@ -0,0 +1,225 @@
+1. kube-apiserver picks up the certificates and start serving them on the configured SNI.
+
+### Topology Considerations
+#### Hypershift / Hosted Control Planes
+N/A
+
+#### Standalone Clusters
+N/A
+
+#### Single-node Deployments or MicroShift
+Enhancement is solely intended for MicroShift.
+
+### Implementation Details/Notes/Constraints
+The certs will be prepended to the []configv1.NamedCertificate list before the api server is started. (it will be added to `-tls-sni-cert-key` flag)
+
+This certificate paths configuration and names will be prepended into the kube-apiserver `tls-sni-cert-key` command line flag.
+When the same SNI appears in the SAN part of the provided certs, this certificate will take precedence over the `default` external-signer. [ref](https://github.com/kubernetes/kubernetes/blob/98358b8ce11b0c1878ae7aa1482668cb7a0b0e23/staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/named_certificates.go#L38)
+
+
+### API Extensions
+N/A
+
+### Risks and Mitigations
+* User provide certificate and it expired after some time.
+  kube-api server will continue serving with an expired cert - similiar approach is taken by OpenShift.
+  > users can mitigate this by using --insecure-skip-tls-verify client mode
+
+* User provided expired Certificate and Microshift service started/restarted 
+  Microshift will start with a warning in the logs, kube-apiserver will continue serve with an expired cert - similiar approach is taken by OpenShift.
+  > users can mitigate this by using --insecure-skip-tls-verify client mode
+
+* Users might configure certs with the wrong names that doesnt match the certificate SAN,but openShift allows it so we will too.
+### Drawbacks
+* External certificate wont be automaticly renewed, so it requires manual Certificate rotation. 
+  similiar approach is taken by OpenShift.
+
+
+## Design Details
+N/A
+
+## Open Questions [optional]
+N/A
+
+## Test Plan
+add e2e test that will:
+- generate and sign certificates with custom ca.
+- change the default Microshift configuration to use the newly generated certs.
+- make sure system is functional using  generated external kubeconfig.
+- intentionally configure invalid value in `names` (ie: 127.0.0.1,localhost) , Microshift should ignore this configuration and skip this certificate.
+
+## Graduation Criteria
+### Dev Preview -> Tech Preview
+- Gather feedback from early adopters on possible issues.
+
+### Tech Preview -> GA
+- Extensive testing with various certificates variations.
+- User facing documentation created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+### Removing a deprecated feature
+N/A
+
+## Upgrade / Downgrade Strategy
+N/A
+
+## Version Skew Strategy
+N/A
+
+## Operational Aspects of API Extensions
+N/A
+
+### Failure Modes
+The provided certs value will be validated before is it passed to the api-server flag
```

Comment:
IIUC, cert and key files passed to `--tls-sni-cert-key` are watched and reloaded dynamically (via https://github.com/openshift/kubernetes/blob/352677df596e5dab4098898974e36df8bcd067e2/staging/src/k8s.io/apiserver/pkg/server/options/serving.go#L323). Is it worthwhile to perform the proposed validation once at startup?

OCP performs some validation before a configured cert and name reach that flag, but in that case the files being passed to the kube-apiserver process are operator-managed as opposed to being directly managed by the user, as in the Microshift case.

- https://github.com/openshift/kubernetes/blob/c3ad486c907cdf30ee97c9a7e7052a823a49c34b/openshift-kube-apiserver/admission/customresourcevalidation/apiserver/validate_apiserver.go#L54
- https://github.com/openshift/cluster-kube-apiserver-operator/blob/e6cba0a210714e935ac20c95873cc63e1175bb43/pkg/operator/configobservation/apiserver/observe_apiserver.go#L91

Reply:
That validation to ensure the settings in the cert don't collide with the internal IPs of the cluster look valuable to avoid breaking the ability of pods to talk to the API server. Thanks for pointing those out, we had a hard time finding anything like that!

Reply:
added table with reserved IP/DNS values  which will be  cause the certificate to be ignored.

Reply:
Do you plan to do anything to prevent the files at `certPath`/`keyPath` passing startup-time validation, being modified in such a way that they would fail startup-time validation, and then being dynamically reloaded at runtime? It's potentially surprising that changing `namedCertificates` in the config file has no effect until the next restart, but changes to the contents of files referenced in the config will be dynamically reloaded (unless I am misunderstanding how the `--tls-sni-cert-key` option works).

Reply:
That does sound like a potentially bad hole that bypasses important validation.